File: AccuPay/Domain/Commons/PayratesCalendar.vb
Option Strict On

Imports AccuPay.Data

Public Class PayratesCalendar

    Private ReadOnly _payrates As IDictionary(Of Date, IPayrate)

    Public Sub New(payrates As IEnumerable(Of IPayrate))
        _payrates = payrates.ToDictionary(Function(p) p.Date)
    End Sub

    Public Function Find([date] As Date) As IPayrate
        If _payrates.ContainsKey([date]) Then
            Return _payrates([date])
        End If

        Return Nothing
    End Function

    Public Function LegalHolidays() As IList(Of IPayrate)
        Return _payrates.
            Where(Function(p) p.Value.IsRegularHoliday).
            Select(Function(p) p.Value).
            ToList()
    End Function

End Class

File: Accupay.Data/Services/PayratesCalendar.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace AccuPay.Data.Services
{
    public class PayratesCalendar
    {
        private readonly IDictionary<DateTime, IPayrate> _payrates;

        public PayratesCalendar(IEnumerable<IPayrate> payrates)
        {
            _payrates = payrates.ToDictionary(p => p.Date);
        }

        public IPayrate Find(DateTime date)
        {
            if (_payrates.ContainsKey(date))
                return _payrates[date];

            return null;
        }

        public IList<IPayrate> LegalHolidays()
        {
            return _payrates.Where(p => p.Value.IsRegularHoliday).Select(p => p.Value).ToList();
        }
    }
}